code/modules/admin/spawn_panel/admin_panels.dm

```diff
@@ -0,0 +1,367 @@
+				to_chat(usr, span_notice("Marked object: [icon2html(target, usr)] [span_bold("[target]")]"))
+				toggle_precise_mode(PRECISE_MODE_OFF)
+				SStgui.update_uis(src)
+
+			if(PRECISE_MODE_COPY)
+				var/copied_type = target.type
+				to_chat(usr, span_notice("Picked object: [icon2html(target, usr)] [span_bold("[target]")]"))
+				src.copied_type = "[copied_type]"
+				toggle_precise_mode(PRECISE_MODE_OFF)
+				SStgui.update_uis(src)
+
+		return TRUE
+
+/datum/admins/spawnpanel/ui_data(mob/user)
+	var/data = list()
+	data["icon"] = selected_object_icon
+	data["iconState"] = selected_object_icon_state
+	data["precise_mode"] = precise_mode
+	data["selected_object"] = selected_object
+	data["copied_type"] = copied_type
+	return data;
+
+/datum/admins/spawnpanel/ui_assets(mob/user)
+	return list(
+		get_asset_datum(/datum/asset/json/spawnpanel),
+	)
+
+/datum/admins/spawnpanel/proc/spawn_item(list/spawn_params)
+	if(!check_rights_for(user_client, R_ADMIN) || !spawn_params)
+		return
+
+	var/path = text2path(spawn_params["object_list"]) || null
+
+	if(!path || (!ispath(path, /obj) && !ispath(path, /turf) && !ispath(path, /mob)))
+		return
+
+	var/amount = clamp(text2num(spawn_params["object_count"]), 1, ADMIN_SPAWN_CAP)
+
+	var/offset_raw = spawn_params["offset"]
+	var/list/offset = splittext(offset_raw, ",")
+	var/X = 0
+	var/Y = 0
+	var/Z = 0
+
+	if(spawn_params["X"] && spawn_params["Y"] && spawn_params["Z"])
+		X = spawn_params["X"]
+		Y = spawn_params["Y"]
+		Z = spawn_params["Z"]
+	else if(offset.len > 0)
+		X = text2num(offset[1])
+		if(isnull(X))
+			X = 0
+
+		if(offset.len > 1)
+			Y = text2num(offset[2])
+			if(isnull(Y))
+				Y = 0
+
+		if(offset.len > 2)
+			Z = text2num(offset[3])
+			if(isnull(Z))
+				Z = 0
+
+	var/obj_dir = text2num(spawn_params["object_dir"]) || 1
+	var/atom_name = sanitize(spawn_params["object_name"])
+	var/where = spawn_params["object_where"]
+	var/atom/target
+
+	if(where == WHERE_MOB_HAND || where == WHERE_TARGETED_MOB_HAND)
+		var/atom/target_reference
+		switch(where)
+			if(WHERE_TARGETED_MOB_HAND)
+				target_reference = spawn_params["object_reference"]
+
+			if(WHERE_MOB_HAND)
+				target_reference = usr
+
+		if(!target_reference)
+			to_chat(usr, span_warning("No target reference provided."))
+			return
+
+		if(!ismob(target_reference))
+			to_chat(usr, span_warning("The targeted atom is not a mob."))
+			return
+
+		if(!iscarbon(target_reference) && !iscyborg(target_reference))
+			to_chat(usr, span_warning("Can only spawn in hand when the target is a carbon mob or cyborg."))
+			where = WHERE_FLOOR_BELOW_MOB
+		target = target_reference
+
+	else if(where == WHERE_MARKED_OBJECT)
+		if(!usr.client.holder.marked_datum)
+			to_chat(usr, span_warning("You don't have any object marked."))
+			return
+		else if(!istype(usr.client.holder.marked_datum, /atom))
+			to_chat(usr, "The object you have marked cannot be used as a target. Target must be of type /atom.", confidential = TRUE)
+			return
```

Comment:
Make these tochats consistent. One uses span warning, one doesn't and is `confidential = TRUE`.